feat(smb): add volume isolation and stage/unstage support to SMB CSI …

MattPOlson opened the pull request:
…driver

- Enables NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME and ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME
- Appends SHA-256 hash of SMB credentials to volume ID for credential-based volume isolation
- NodePublishVolume updated to bind-mount from staging path to pod volume path
- Implements NodeUnstageVolume with reference count check using /proc/mounts
- Preserves all existing SMB CSI functionality (Kerberos, GID, subDir handling)



**What type of PR is this?**
/kind feature
/kind design


**What this PR does / why we need it**:
This PR enhances the SMB CSI driver to support STAGE_UNSTAGE_VOLUME, enabling a single SMB share mount per node with bind mounts for each pod. It also introduces credential-based volume ID hashing to isolate mounts for identical shares accessed with different credentials.

**Which issue(s) this PR fixes**:
Fixes #353 
Fixes  #935 

Fixes #

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
- [x] includes documentation
- [ ] adds unit tests
- [x] tested upgrade from previous version

**Special notes for your reviewer**:
This is a non-breaking change that enhances driver correctness and security without removing or deprecating existing functionality. Happy to add follow-up documentation or test case references.

**Release note**:
```
none
```
[testresults.txt](https://github.com/user-attachments/files/19799698/testresults.txt)

gnufied commented:
/ok-to-test

andyzhangx reviewed (commented):
@MattPOlson thanks for the fix, some comments:
> Adds credential-based hashing to the volume ID so PVCs using the same share + same credentials resolve to the same global mount.

from my understanding, this behavior change does not solve the issue since in common scenario one file share using one credentials

> Uses NodeStageVolume + bind mounts to avoid duplicate mounts entirely.

I don't find any code change in `NodeStageVolume` function in your PR

> Updates NodeUnstageVolume to safely unmount the global share only when it's no longer referenced by any bind mount.

I think this is the most important fix that solves this issue! could you only make this change in the PR and we could skip windows related change since Windows does not allow multiple mounts on the same smb share.

Thanks

MattPOlson commented:
@andyzhangx 

Thanks for the thoughtful feedback, really appreciate the detailed review.

You're absolutely right that the NodeUnstageVolume() enhancement is the core fix. That reference-checking logic is what ensures the driver does not prematurely unmount the global SMB share while bind mounts from other pods are still active. Without this, the kubelet can misinterpret active usage and trigger unmount errors, especially when multiple pods or workloads interact with the same SMB share.

Regarding the credential-based hashing, I added that to address a secondary concern mentioned in #353 specifically, the risk of a new PVC unintentionally reusing a stale global mount that was originally mounted with different credentials. While that scenario is rare (most real-world use cases use a single credential per share), I wanted to guard against accidental credential leakage. That said, I’m completely fine with removing this from the PR to keep the scope focused.

On the NodeStageVolume() question confirmed, there were no functional changes made to that method in this PR. All staging behavior remains as it was upstream. I originally had an idea around that but later changed my mind.

The Windows-related changes were added purely to support build compatibility. The reference-checking logic in NodeUnstageVolume() depends on /proc/mounts, which doesn’t exist on Windows. To prevent build and e2e test failures, I added a HasMountReferences() helper function with a Linux implementation and stubbed no-op versions for Windows and Darwin.

In summary, I'm happy to:

Keep this PR focused on just the NodeUnstageVolume() fix (with required platform support)

Remove the credential hashing logic

Let me know if you’d prefer this rebased down to just the essential fix, I can prep that immediately. Thanks again for helping guide this improvement!

andyzhangx commented on pkg/smb/smb_common_linux.go:
```diff
@@ -48,3 +51,23 @@ func prepareStagePath(_ string, _ *mount.SafeFormatAndMount) error {
 	return os.Mkdir(name, perm)
 }
+
+func HasMountReferences(stagingTargetPath string) (bool, error) {
```
can you add explanation for this func? I still doubt how did you solve the problem by only searching `stagingTargetPath` since in my testing environment (with 2 replicas referencing to one PVC), the bind mount would be like following, you need to count the target mount share during `NodeUnstageVolume` to make sure there are no other mounts referencing the smb share.

```
# cat /proc/mounts | grep smb
//smb-server.default.svc.cluster.local/share/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9 /var/lib/kubelet/plugins/kubernetes.io/csi/smb.csi.k8s.io/085ffd65e2835034cdf2a23f67a498673427c7497def5d808bfb505c0df0b1a4/globalmount cifs rw,relatime,vers=3.1.1
//smb-server.default.svc.cluster.local/share/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9 /var/lib/kubelet/pods/4aad1717-a5cb-4f93-a28b-2a8a8c36d1bd/volumes/kubernetes.io~csi/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9/mount cifs rw,relatime,vers=3.1.1
//smb-server.default.svc.cluster.local/share/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9 /var/lib/kubelet/pods/2164da99-d68f-4886-9bcb-9bb9a42c844f/volumes/kubernetes.io~csi/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9/mount cifs rw,relatime,vers=3.1.1
```

MattPOlson replied:
> can you add explanation for this func? I still doubt how did you solve the problem by only searching `stagingTargetPath` since in my testing environment (with 2 replicas referencing to one PVC), the bind mount would be like following, you need to count the target mount share during `NodeUnstageVolume` to make sure there are no other mounts referencing the smb share.
> 
> ```
> # cat /proc/mounts | grep smb
> //smb-server.default.svc.cluster.local/share/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9 /var/lib/kubelet/plugins/kubernetes.io/csi/smb.csi.k8s.io/085ffd65e2835034cdf2a23f67a498673427c7497def5d808bfb505c0df0b1a4/globalmount cifs rw,relatime,vers=3.1.1
> //smb-server.default.svc.cluster.local/share/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9 /var/lib/kubelet/pods/4aad1717-a5cb-4f93-a28b-2a8a8c36d1bd/volumes/kubernetes.io~csi/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9/mount cifs rw,relatime,vers=3.1.1
> //smb-server.default.svc.cluster.local/share/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9 /var/lib/kubelet/pods/2164da99-d68f-4886-9bcb-9bb9a42c844f/volumes/kubernetes.io~csi/pvc-00a18b4a-977d-4ce8-a911-fc661a7332f9/mount cifs rw,relatime,vers=3.1.1
> ```

You’re right that our initial implementation of HasMountReferences() was scanning /proc/mounts for paths prefixed by stagingTargetPath, which doesn't fully reflect how bind mounts work in real-world deployments. As you pointed out, kubelet performs bind mounts from the global path into pod-specific paths like /var/lib/kubelet/pods/.../volumes/.../mount, and these paths are not subdirectories of the staging path.

The correct approach here is to:

Parse /proc/mounts

Count the number of entries where the source (the SMB share URI) matches the mounted share

Only proceed with unmounting if stagingTargetPath is the last remaining mount target

I’ll update the implementation of HasMountReferences() to reflect this, by comparing entries that have the same mount source as the staging path. This aligns with how GetDeviceMountRefs() works internally in kubelet, and ensures that the global mount is only unmounted once all bind mounts are gone.

Thanks for flagging this — I’ll update the PR accordingly.

andyzhangx replied:
@MattPOlson thanks for the work, pls also provide the /proc/mounts examples when there are multiple PVCs using the same file share in the PR description, I think we only need this fix right now, thanks.

andyzhangx commented:
@MattPOlson thanks for the detailed explanations, pls only make necessary changes to fix the original issue.

MattPOlson commented:
> @MattPOlson thanks for the detailed explanations, pls only make necessary changes to fix the original issue.

@andyzhangx 

Thanks again for the review, I’d like to clarify why the volumeHandle normalization is essential to fully solving the problem we're seeing in production, even with the upstream behavior.

**How This Happens with the Current Upstream Driver**
Although the upstream SMB CSI driver generates a unique volumeHandle per PVC, those handles can still reference the same underlying SMB share (e.g., //smb-server/share). This leads to the following behavior:

Multiple PVCs, each with unique volumeHandles, stage the same SMB share.

This results in multiple global mount paths, such as:

/plugins/.../<vol1>/globalmount

/plugins/.../<vol2>/globalmount

But from the Linux kernel's perspective, all those mounts reference the same CIFS device.

When one PVC is deleted, the kubelet tries to unmount its corresponding globalmount path.

However, GetDeviceMountRefs() detects that the same source is still mounted in other locations, and kubelet refuses to unmount, thinking it's still in use.

We've confirmed this with logs like:


`GetDeviceMountRefs check failed ... device mount path is still mounted by other references [...]`

This behavior becomes more common when Helm charts, templates, or automation generate many PVCs that point to the same SMB share.

 **How volumeHandle Normalization Solves This**
By reintroducing a deterministic volumeHandle, computed as a hash of source + credentials, we can ensure:

The same share + creds → same volumeHandle

Only one global mount is created per node

Additional PVCs referencing the same share use bind mounts only (via NodePublishVolume)

NodeUnstageVolume only unmounts when no other pods are using the mount

This approach has several advantages:

-  Prevents GetDeviceMountRefs() collisions and premature unmount failures
-  Reduces duplicate mounts and simplifies cleanup
-  Enforces credential-based isolation (if creds differ, mount paths differ)
-  Retains existing upstream logic for all other paths

 **TL;DR**
The current upstream behavior causes multiple global mounts to the same share

Linux sees them as the same device and blocks unmounts

Our changes resolve this by ensuring the share is only mounted once per node

The fix is safe, backward-compatible, and improves both reliability and security

Happy to reintroduce this logic in a scoped commit if you'd like to review it in isolation — thanks again for considering!

MattPOlson commented:
Hi @andyzhangx  and maintainers,

 I'd like to request a status update on this PR. It's been marked as "rotten" due to inactivity, but I've made some improvements to address the feedback and concerns raised during the review process.

  Recent Changes Made to Improve the PR:
  
 1. Enhanced Security and Reliability (commit d331a8db7)

  - Upgraded hash collision resistance from 64-bit to 128-bit using Base64 encoding to address enterprise security concerns
  - Implemented atomic double-check pattern to mitigate race conditions during mount reference checks
  - Added robust mount path parsing with proper subdirectory validation to prevent false positives
  - Included pre-unmount safety checks to prevent race conditions during NodeUnstageVolume
  
  2. CSI Specification Compliance (commit 0c6a188da)

  - Removed PUBLISH_UNPUBLISH_VOLUME capability since SMB shares are mounted directly by nodes
  - Fixed sanity test suite failures by not advertising unsupported controller functionality
  - Maintained full node-side STAGE_UNSTAGE_VOLUME support which is the core functionality needed
  
I believe this PR now addresses all the concerns raised and would greatly benefit users experiencing mount reference issues like we currently are.

Thanks!
